front/lib/actions/helpers.ts

```diff
@@ -0,0 +1,104 @@
+import {
+  ActionResponseBaseSchema,
+  isActionResponseBase,
+} from "@app/lib/actions/types";
+import { Authenticator, prodAPICredentialsForOwner } from "@app/lib/auth";
+import {
+  DustAPI,
+  DustAPIErrorResponse,
+  DustAppConfigType,
+} from "@app/lib/dust_api";
+import { Err, Ok, Result } from "@app/lib/result";
+
+interface CallActionParams<V extends t.Mixed> {
+  workspaceId: string;
+  input: { [key: string]: unknown };
+  action: Action;
+  config: DustAppConfigType;
+  // needs to be an io-ts schema of the value returned by the action
+  // Dust API response is {results: [[{value: V}]]}
+  responseValueSchema: V;
+}
+
+/**
+ * This function is **not** intended to be used by the client directly.
+ *
+ * It is used server-side to call an action on the production API, when streaming is not required.
+ * It has the advantage of providing an interface that validates the response of the action using io-ts.
+ *
+ * @param workspaceId string the workspace id (sId)
+ * @param input { [key: string]: unknown } the action input (a single input)
+ * @param config DustAppConfigType the action config
+ * @param responseValueSchema V extends t.Mixed the io-ts schema of the action response value
+ */
+export async function callAction<V extends t.Mixed>({
+  workspaceId,
+  input,
+  action,
+  config,
+  responseValueSchema,
+}: CallActionParams<V>): Promise<
+  Result<t.TypeOf<typeof responseValueSchema>, DustAPIErrorResponse>
+> {
+  const app = cloneBaseConfig(action.app);
+
+  const owner = (
+    await Authenticator.internalBuilderForWorkspace(workspaceId)
+  ).workspace();
+  if (!owner) {
+    throw new Error(
+      `Could not get internal builder for workspace ${workspaceId}`
+    );
+  }
+  const prodCredentials = await prodAPICredentialsForOwner(owner);
+
+  const prodAPI = new DustAPI(prodCredentials);
+
+  const response = (await prodAPI.runApp(app, config, [input])) as Result<
+    unknown,
+    DustAPIErrorResponse
+  >;
+
+  if (response.isErr()) {
+    return response;
+  }
+
+  // create a schema validator using the provided schema + the base response schema
+  const responseSchema = t.intersection([
+    ActionResponseBaseSchema,
+    t.type({
+      results: t.array(t.array(t.type({ value: responseValueSchema }))),
+    }),
+  ]);
+  type responseType = t.TypeOf<typeof responseSchema>;
+  const responseChecker = (response: unknown): response is responseType =>
+    isRight(responseSchema.decode(response));
+
+  if (responseChecker(response.value)) {
+    // the response is a valid success response for the action
+    // return the "value" field of the first result
+    return new Ok(response.value.results[0][0].value);
+  }
+
+  if (isActionResponseBase(response.value)) {
+    // the response is of the right shape, but it's not a success response
+    return new Err({
+      type: "action_failed",
+      message: `Doc Tracker action failed response: ${JSON.stringify(
+        response.value.status
```

Comment:
This is likely way tooo big ? (can be Mb)

Reply:
Makes sense, for this one maybe I should dumb the keys + statuses ?

Reply:
Yes anything what does not have the traces or results :)

Reply:
How can this be megabytes ? It's not the trace, just the status for each block no ? 
<img width="490" alt="Screenshot 2023-07-17 at 15 25 01" src="https://github.com/dust-tt/dust/assets/14199823/99a4c710-6e20-4fd4-9333-03e2192bb690">

should be this shape if I'm not mistaken ?

Reply:
Ah sorry I missed that it was response.value.status thought it was response.value. reponse.value.status is fine :+1: